File: algorithmic_toolbox/week5_dynamic_programming_1/1_money_change_again/change_dp.cpp
#include <algorithm>
#include <iostream>
#include <vector>

const std::vector<int> COINS = {1, 3, 4};

int get_change(int m) {
  std::vector<int> dp(m + 1, m + 1);

  for (int i = 1; i <= m; i++) {
    if (find(COINS.begin(), COINS.end(), i) != COINS.end())
      dp[i] = 1;
    else
      for (const auto& coin : COINS)
        if (i >= coin)
          dp[i] = std::min(dp[i], dp[i - coin] + 1);
  }

  if (dp[m] > m)
    throw std::runtime_error("Something went wrong");

  return dp[m];
}

int main() {
  int m;
  std::cin >> m;
  std::cout << get_change(m) << std::endl;

  return 0;
}

File: algorithmic_toolbox/week5_dynamic_programming_1/1_money_change_again/change_dp.py
import sys

COINS = (1, 3, 4)


def get_change(m: int) -> int:
    dp = [m + 1] * (m + 1)

    for i in range(1, m + 1):
        dp[i] = (
            1 if i in COINS else min((dp[int(j)] for j in (i - coin for coin in COINS) if int(j) == j)) + 1  # noqa: FURB123
        )

    if dp[m] > m:
        raise Exception("something went wrong")

    return dp[m]


if __name__ == "__main__":
    m = int(sys.stdin.read())
    print(get_change(m))
